### PR TITLE
separated Diffusion Model resampler for STM PR from PR #1794

### PR DESCRIPTION
**Pull request by YongyiBWu**
Sub-PR with only diffusion model resampler update, separated from PR #1794 
Suggested changes from PR#1794 were applied. Response details see PR #1794

=================================
This pull request introduces a new score-based diffusion model machine learning tool, integrates it into the build system, and adds new modules for virtual detector resampling using this model. Additionally, it updates the generator ID enumeration to support a new generator and defines transformation defaults for the VDResampler. The changes are grouped below by theme.

**Machine Learning Model Integration:**

* Added a new `ScoreBasedDiffusionModel` class for training and sampling using score-based diffusion models, including its header and implementation. This model supports configurable neural network architectures, optimizer and noise schedule selection, and model persistence. (`MachineLearningTools/inc/ScoreBasedDiffusionModel.hh`, `MachineLearningTools/CMakeLists.txt`, `MachineLearningTools/src/SConscript`) [[1]](diffhunk://#diff-073c8ac79f88d459840531da77345137c1a906a86cf1df8e734ad912afa79b33R1-R321) [[2]](diffhunk://#diff-205ccf97771e4c8eb087aa89d40a6b27c0d1b15917ec3d912ee5e45075657bb4R1-R11) [[3]](diffhunk://#diff-73024793e9b198b17c3b786c95f4585f5af847f6d2a642f6d33489e91b7b3d0fR1-R20)
* Integrated the new machine learning library into the build system and made it available to other modules. (`MachineLearningTools/CMakeLists.txt`, `MachineLearningTools/src/SConscript`, `STMMC/src/SConscript`) [[1]](diffhunk://#diff-205ccf97771e4c8eb087aa89d40a6b27c0d1b15917ec3d912ee5e45075657bb4R1-R11) [[2]](diffhunk://#diff-73024793e9b198b17c3b786c95f4585f5af847f6d2a642f6d33489e91b7b3d0fR1-R20) [[3]](diffhunk://#diff-3ecc2ddbf5e410a13f71b7bb6e3d98a7e612a31ac83551a8c67bf56c181d8339R51-R52)

**Virtual Detector Resampling Modules:**

* Added four new art modules for VDResampler: configuration, training, generation from model, and generation with mixing. These modules utilize the new machine learning tools and are registered in the build system. (`STMMC/CMakeLists.txt`)

**Generator ID Updates:**

* Extended the `GenId` enumeration and its string mapping to include a new generator, `STMDownStreamGenTool`. (`MCDataProducts/inc/GenId.hh`) [[1]](diffhunk://#diff-ba83e82733fbd19dce7f24a2661ea3eb9e285e3e8d1d4c22ccf549dad2fbd708L47-R48) [[2]](diffhunk://#diff-ba83e82733fbd19dce7f24a2661ea3eb9e285e3e8d1d4c22ccf549dad2fbd708L67-R67)

**VDResampler Transformation Defaults:**

* Added a header defining default transformation constants for the VDResampler. (`STMMC/inc/VDResamplerTransformDefaults.hh`)

### CODE REVIEW

**Comment by oksuzian** [comment]
I'll review this pull request by examining the details and the code changes.Let me fetch the changed files directly from the PR's head branch to review them.Now let me also check the existing review comments on this PR:Here is my review of [PR #1798](https://github.com/Mu2e/Offline/pull/1798). This is a substantial PR (~2,967 additions) introducing a custom score-based diffusion model ML framework and four new art modules for virtual detector resampling. I've categorized the issues by severity.

---

## 🔴 Bugs / Correctness Issues

### 1. `HepLorentzVector` constructor argument order is wrong
In both `VDResamplerGenerateFromModel_module.cc` and `VDResamplerGenerateMix_module.cc`, the `HepLorentzVector` is constructed as:
```cpp name=STMMC/src/VDResamplerGenerateFromModel_module.cc url=https://github.com/Mu2e/Offline/pull/1798/files#diff-2
const CLHEP::HepLorentzVector fourMomParticle(eTotal, momParticle);
```
CLHEP's `HepLorentzVector(double, Hep3Vector)` interprets the first argument as the **x-component**, not energy. The correct constructor for (energy, 3-momentum) is `HepLorentzVector(Hep3Vector, double)`:
```cpp
const CLHEP::HepLorentzVector fourMomParticle(momParticle, eTotal);
```
This will produce completely wrong 4-vectors as written.

### 2. `loadModel()` double-initializes weights wastefully and risks inconsistency
In `ScoreBasedDiffusionModel::loadModel()`, the constructor is called (which randomly initializes all weights), and then immediately overwritten:
```cpp name=MachineLearningTools/src/ScoreBasedDiffusionModel.cc url=https://github.com/Mu2e/Offline/pull/1798/files#diff-3
ScoreBasedDiffusionModel model(randFlat, randGaussQ, dim, ...);
// Overwrite initialized weights with loaded values
for (size_t l = 0; l < model.network_.size(); ++l) {
    model.network_[l].W = loadedNetwork[l].W;
    model.network_[l].b = loadedNetwork[l].b;
}
```
This wastes random engine state (consuming random numbers during construction), which subtly changes reproducibility compared to a model that was never loaded. Consider adding a private constructor that skips random initialization.

### 3. `trainingSampleSize_` assigned from `size_t` to `int` — truncation
In `train()`, `trainingSampleSize_ = N;` where `N` is `size_t` but `trainingSampleSize_` is `int`. For very large datasets (>2^31 samples), this silently overflows.

### 4. `VDResamplerConfigure` — `ttree` is uninitialized when `doROOTDump` is false
The member `TTree* ttree;` is never initialized to `nullptr`. If `doROOTDump` is `false`, `ttree` is a dangling pointer. While it's not dereferenced in that path currently, this is a latent bug.

---

## 🟡 Robustness / Defensive Programming Issues

### 5. No validation of `VDr > 0` in training or generation modules
As also noted in [an existing review comment](https://github.com/Mu2e/Offline/pull/1798#discussion_r3068701672), if `VDr` is configured as 0 or negative, the division `rho = r / VDr` produces inf/NaN. Both `VDResamplerTrain`, `VDResamplerGenerateFromModel`, and `VDResamplerGenerateMix` should validate `VDr > 0` in their constructors.

### 6. Silent fallback for unknown optimizer/noise schedule strings
In `VDResamplerTrain_module.cc`, unrecognized `SBDMoptimizer` or `SBDMnoiseSchedule` values silently default to ADAM/COSINE. A typo like `"Aadam"` would silently use the wrong optimizer. Add validation or throw on unknown values.

### 7. Fisher-Yates shuffle has off-by-one potential
```cpp name=MachineLearningTools/src/ScoreBasedDiffusionModel.cc
size_t j = static_cast<size_t>(randFlat_.fire() * (i + 1));
```
`RandFlat::fire()` returns `[0, 1)`, so `j` ranges over `[0, i]` which is correct. However, if the engine ever returns exactly 1.0 (which some implementations may do), `j == i+1` would be out of bounds. A safer approach: `j = std::min(j, i)`.

### 8. `computeLoss()` is declared but never called
The member function `computeLoss()` is implemented but the loss is actually computed inline in the training loop. This is dead code that could become stale.

---

## 🟢 Style / Maintenance Issues

### 9. Inconsistent coding style
The codebase mixes different formatting conventions throughout:
- Sometimes `for(size_t i=0;i<z.size();++i)` (no spaces), other times `for (size_t i = 0; i < out; ++i)` (with spaces)
- Inconsistent brace placement (`{` on same line vs. next line)
- The Mu2e Offline codebase generally uses consistent formatting — this should be unified.

### 10. `typedef` instead of `using`
```cpp
typedef cet::map_vector_key key_type;
typedef unsigned long VolumeId_type;
```
Modern C++ prefers `using` declarations. Also, `key_type` is defined but never used in some modules.

### 11. Member variables used as scratch space in `analyze()`
In `VDResamplerConfigure` and `VDResamplerTrain`, variables like `x`, `y`, `z`, `px`, `py`, `pz`, `pdgId` are stored as class members but are only used as local variables within `analyze()`. They should be local variables instead — storing them as members wastes memory and makes the class harder to reason about thread-safety.

### 12. Duplicated inverse-transform code in Generate modules
The coordinate inverse transform (arctanh, polar→Cartesian, momentum recovery) is duplicated verbatim between `VDResamplerGenerateFromModel_module.cc` and `VDResamplerGenerateMix_module.cc`. This should be extracted into a shared utility function (e.g., in a header under `STMMC/inc/`).

### 13. Magic numbers
Several magic numbers appear without constants:
- `100000` threshold for two-stage training in `VDResamplerConfigure`
- `1e-6` for radius safety check in multiple files
- `0.1` for minimum time in `t_safe = std::max(time, 0.1)`

These should be named constants or configurable parameters.

### 14. No `#include <memory>` in modules using `std::unique_ptr`
Several module source files use `std::unique_ptr` without explicitly including `<memory>`. This compiles because other headers transitively include it, but it's fragile.

---

## Summary

| Severity | Count | Key Items |
|----------|-------|-----------|
| 🔴 Bug | 4 | Wrong `HepLorentzVector` arg order, uninitialized `ttree`, int truncation, RNG state pollution |
| 🟡 Robustness | 4 | No `VDr` validation, silent optimizer fallback, shuffle edge case, dead code |
| 🟢 Style | 6 | Duplicated transforms, member-as-local, magic numbers, inconsistent formatting |

The most critical fix needed is the **`HepLorentzVector` constructor argument order** — this will produce wrong physics results silently.

File: MachineLearningTools/src/ScoreBasedDiffusionModel.cc [hunk 1]
```diff
@@ -0,0 +1,1048 @@
...
+            gradientClipThreshold,
+            learningRate,
+            diffusionSteps
+        );
+
+        // Overwrite initialized weights with loaded values
+        for (size_t l = 0; l < model.network_.size(); ++l) {
+            model.network_[l].W = loadedNetwork[l].W;
+            model.network_[l].b = loadedNetwork[l].b;
+        }
```
[the remark this answers is not in the document]
Comment (YongyiBWu) [reply MachineLearningTools/src/ScoreBasedDiffusionModel.cc]: Addressed in the new commit.

File: STMMC/src/VDResamplerTrain_module.cc [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+    ScoreBasedDiffusionModel::OptimizerType opt =
+        (conf().SBDMoptimizer() == "SGD") ?
+        ScoreBasedDiffusionModel::OptimizerType::SGD :
+        ScoreBasedDiffusionModel::OptimizerType::ADAM;
+
+    // noise schedule
+    ScoreBasedDiffusionModel::NoiseScheduleType sched =
+        (conf().SBDMnoiseSchedule() == "LINEAR") ?
+        ScoreBasedDiffusionModel::NoiseScheduleType::LINEAR :
+        ScoreBasedDiffusionModel::NoiseScheduleType::COSINE;
```
[the remark this answers is not in the document]
Comment (YongyiBWu) [reply STMMC/src/VDResamplerTrain_module.cc]: Added in the new commit.

File: STMMC/src/VDResamplerTrain_module.cc [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+
+      // shift to detector-centered coordinates
+      double dx = x_extrapolated - x0;
+      double dy = y_extrapolated - y0;
+      // polar coordinates
+      double r = std::sqrt(dx*dx + dy*dy);
+      double rho = r / VDr;
+      // numerical safety (avoid rho >= 1)
+      const double eps = 1e-6;
+      rho = std::min(rho, 1.0 - eps);
```
[the remark this answers is not in the document]
Comment (YongyiBWu) [reply STMMC/src/VDResamplerTrain_module.cc]: Additional checks added in the new commit.

**Comment by YongyiBWu** [comment]
1. Geant4.10 and 11 both have HepLorentzVector(double, Hep3Vector) constructors according to their documentations, and some existing modules us that too. But to make things more consistent throughout the repo, all changed to HepLorentzVector(Hep3Vector, double) constructor now.
2. Since the loaded model is never used for retrain and only used for generating particles, and the model size remains the same, the reproducibility is not a problem. Random states may be wasted but not a large problem. Still, I added an optional flag to the constructor that turns off the RNG initialization when loading from a saved model. 
3. trainingSampleSize_ has been changed to size_t. Comment added to remind use of std::stoull if parsing from a saved file.
4. Added a nullptr initialization.
5. VDr, VDz0 values are now validated in all 3 modules.
6. Now unrecognized settings prompt a warning message before the values fall back  to default.
7. This does not really matter. The RandFlat.fire() returns in [0,1) so there should not be a problem with j=i+1. But I guess it doesn't hurt to add a line j = std::min(j, i);
8. Now the loss per dimension calculation calls the computeLoss function instead of using its own implementation. Both loss and gradient calculations now unite to a per dimensional basis, and additional dimensional check were added in computeLoss.
9. ... Fine... spacing conventions unified.
10. key_type were removed. "typedef unsigned long VolumeId_type;" also used in multiple other files. Kept for consistency.
11. These are used to store variables and potentially used for writing into ROOT dumps. Not making sense to move them to local vars.
12. Transform methods along with relevant constants are now collected in a single header file VDResamplerTransforms.hh
13. Added in corresponding header file or declared as a constant.
14. Fixed as suggested.

**Comment by YongyiBWu** [comment]
@FNALbuild run build test

**Review by brownd1978 — APPROVED** [review]
I can't evaluate this code technically. Since it's a pure addition I think its OK.
